### PR TITLE
Add unit tests for admission handlers

### LINKED ISSUE #2854: Add unit tests for Gardener Admission Controller
**How to categorize this issue?**

/area quality
/kind task
/priority normal

**What would you like to be added**:
Add unit tests for the namespace deletion and kubeconfig check:
https://github.com/gardener/gardener/tree/master/pkg/admissioncontroller/server/handlers/webhooks

**Why is this needed**:
These code branches are untested at the moment.

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area quality
/kind cleanup
/kind enhancement
/priority normal

**What this PR does / why we need it**:
This PR refactors the Gardener admission webhooks and adds unit tests for the different handlers.

**Which issue(s) this PR fixes**:
Fixes #2854

**Special notes for your reviewer**:

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

File: pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+						p := project(namespace.Name, "project2")
+						t := metav1.Now()
+						p.SetDeletionTimestamp(&t)
+						return p
+					}(),
+				).
+				withNamespace(*namespace).
+				withShoots(client.InNamespace(namespace.Name), shoot("namespace1", "shoot1")),
+			false, "there are still Shoots"),
+		Entry("should not pass because namespace is used by namespace", namespace,
```
Comment (ialidzhikov) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go]:
```suggestion
		Entry("should not pass because namespace is used by project", namespace,
```

File: pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+			newMockClient().
+				withProjects(
+					project("namespace1", "project1"),
+					project(namespace.Name, "project2"),
+				).
+				withNamespace(*namespace),
+			false, "you must delete the corresponding project \"project2\""),
+		Entry("should pass because namespace is used by namespace but not found by client", namespace,
```
Comment (ialidzhikov) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go]:
```suggestion
		Entry("should pass because namespace is used by project but not found by client", namespace,
```

File: pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+						p := project(namespace.Name, "project2")
+						t := metav1.Now()
+						p.SetDeletionTimestamp(&t)
+						return p
+					}(),
+				).
+				withNamespace(*namespace).
+				withShoots(client.InNamespace(namespace.Name)),
+			true, emptyMessage),
+		Entry("should not pass because namespace is used by namespace w/ deletionTimestamp but has shoots", namespace,
```
Comment (ialidzhikov) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go]:
```suggestion
		Entry("should not pass because namespace is used by project w/ deletionTimestamp but has shoots", namespace,
```

File: pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+		Entry("should pass because no project or shoots available", namespace,
+			newMockClient().withProjects(),
+			true, emptyMessage),
+		Entry("should pass because namespace is not project related", namespace,
+			newMockClient().
+				withProjects(
+					project("namespace1", "project1"),
+					project("namespace2", "project2"),
+				), true, emptyMessage),
+		Entry("should pass because namespace is used by namespace w/ deletionTimestamp", namespace,
```
Comment (ialidzhikov) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go]:
```suggestion
		Entry("should pass because namespace is used by project w/ deletionTimestamp", namespace,
```

**Review by ialidzhikov — COMMENTED** [review]
Minor nits

File: pkg/admissioncontroller/server/handlers/webhooks/utils.go [hunk 5]
```diff
@@ -29,6 +29,13 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 )
 
+const (
+	// maxRequestBody is the maximum size of the `AdmissionReview` body.
+	// Same, fixed value from API server for general safety to reduce the odds of OOM issues while reading the body.
+	// https://github.com/kubernetes/kubernetes/blob/d8eac8df28e6b50cd0f5380e23fc57daaf92972e/staging/src/k8s.io/apiserver/pkg/server/config.go#L322
+	maxRequestBody = 3 * 1024 * 1024
```
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/utils.go:36]: Hmm, I know this doesn't change with this PR, but still one question:
I think we are not storing gardener resources in protobuf in etcd. Should we be more careful than k/apiserver here or do you think, we can neglect this?

File: pkg/admissioncontroller/server/handlers/webhooks/utils_test.go [hunk 6]
```diff
@@ -53,15 +53,15 @@ var _ = Describe("Utils tests", func() {
 			scheme  = runtime.NewScheme()
 			decoder = serializer.NewCodecFactory(scheme).UniversalDecoder()
 
-			sizeRequest = int64(336)
+			sizeRequest = int64(342)
```
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/utils_test.go:57]: One side question about `#DecodeAdmissionRequest - should not succeed because of wrong content type`:
What does actually happen, when someone talks to the API server using protobuf? Are webhook requests always send in JSON from the API server?

File: pkg/admissioncontroller/server/handlers/webhooks/validate_kubeconfig_secrets_test.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+			}
+		}
+	)
+
+	DescribeTable("Validate Kubeconfig",
+		func(objFn func() runtime.Object, subresource string, op admissionv1beta1.Operation, expectedAllowed bool, expectedMsg string) {
+			scheme := runtime.NewScheme()
+			Expect(corev1.AddToScheme(scheme)).To(Succeed())
+
+			user := func() authenticationv1.UserInfo {
```
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_kubeconfig_secrets_test.go:140]: Why do you use a func that is invoked a few lines below here instead of a literal?

File: pkg/admissioncontroller/server/handlers/webhooks/validate_kubeconfig_secrets_test.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+					Groups:   []string{"group"},
+				}
+			}
+			response := httptest.NewRecorder()
+			validator := NewValidateKubeconfigSecretsHandler()
+
+			request := createHTTPRequest(objFn(), scheme, user(), op)
+			if subresource != "" {
+				request = createHTTPRequestForSubresource(subresource)
+			}
```
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_kubeconfig_secrets_test.go]: Is this still needed? I think it's not executed anywhere.

File: pkg/admissioncontroller/server/handlers/webhooks/validate_kubeconfig_secrets_test.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+			Expect(admissionReview.Response.Allowed).To(Equal(expectedAllowed))
+			if expectedMsg != "" {
+				Expect(admissionReview.Response.Result).To(
+					PointTo(MatchFields(IgnoreExtras, Fields{
+						"Message": ContainSubstring(expectedMsg),
+					})))
+			}
+
+		},
+		Entry("request is empty (create)", empty, noSubResource, admissionv1beta1.Create, false, "missing admission request"),
```
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_kubeconfig_secrets_test.go]: Can you add a test case for passing an object other than a Secret?

File: pkg/admissioncontroller/server/handlers/webhooks/validate_kubeconfig_secrets_test.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+		Entry("should pass because no Kubeconfig is found (create)", noKubeconfigSecret, noSubResource, admissionv1beta1.Create, true, emptyMessage),
+		Entry("should pass because Kubeconfig is valid (create)", validKubeconfigSecret, noSubResource, admissionv1beta1.Create, true, emptyMessage),
+		Entry("should fail because Kubeconfig is invalid (create)", invalidKubeconfigSecret, noSubResource, admissionv1beta1.Create, false, "exec configurations are not supported"),
```
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_kubeconfig_secrets_test.go]: Can you also add a test case for passing a Secret with `kubeconfig` set to some random string (or just not a valid kubeconfig yaml)?

File: pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+			Expect(admissionReview.Response.Allowed).To(Equal(expectedAllowed))
+			if expectedMsg != "" {
+				Expect(admissionReview.Response.Result).To(
+					PointTo(MatchFields(IgnoreExtras, Fields{
+						"Message": ContainSubstring(expectedMsg),
+					})))
+			}
+
+		},
+		Entry("should pass because no project or shoots available", namespace,
```
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go:103]: Can you add test cases for 
- a different operation than delete
- an empty admission request

just like in the `ValidateKubeconfigSecret` test?

File: pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "foo",
+			},
+		}
+	)
+
+	DescribeTable("Namespace deletion admission",
+		func(namespace *corev1.Namespace, cl client.Client, expectedAllowed bool, expectedMsg string) {
+			ctrl := gomock.NewController(GinkgoT())
```
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go]: I find it a bit confusing, that you create a lot of mocks here, for which calls are never verified. So I assume you only mock the clients in order to have some preexisting objects.

If that's correct and you don't need to check, that explicit calls are made, I would replace the `mockClient` by [fake clients](https://github.com/kubernetes-sigs/controller-runtime/tree/master/pkg/client/fake) and the mock cache and `syncedInformer` by [fake informers](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/cache/informertest/fake_cache.go).
We shouldn't need to reimplement the stuff already available upstream, if it's sufficient for our use cases.

Another option without mocking or faking anything would be to use a local control plane (envtest).

WDYT?

**Review by timebertt — COMMENTED** [review]
Nice PR, thank you very much!
I have mostly nit comments, but also one general comment on the mocks.

File: pkg/admissioncontroller/server/handlers/webhooks/utils.go (hunk 5 above)
Comment (timuthy) [reply pkg/admissioncontroller/server/handlers/webhooks/utils.go:36]: The problem in general is, that you can't be sure about the `--max-request-bytes` configuration of etcd backing the GAPI server. That's the main reason why for the moment I decided to choose the same value as K8s and gain some experience with it. In the worst case, GAPI returns an error because the object cannot be saved in etcd, so even if this value was too high, it wouldn't really harm or cause service outages.

File: pkg/admissioncontroller/server/handlers/webhooks/utils_test.go (hunk 6 above)
Comment (timuthy) [reply pkg/admissioncontroller/server/handlers/webhooks/utils_test.go:57]: > Are webhook requests always send in JSON from the API server?

Concerning my research https://github.com/gardener/gardener/pull/2781#discussion_r488542877, yes.

File: pkg/admissioncontroller/server/handlers/webhooks/utils.go (hunk 5 above)
Comment (timebertt) [reply pkg/admissioncontroller/server/handlers/webhooks/utils.go:36]: Ok, that sounds good. Thanks for explaining.

File: pkg/admissioncontroller/server/handlers/webhooks/utils_test.go (hunk 6 above)
Comment (timebertt) [reply pkg/admissioncontroller/server/handlers/webhooks/utils_test.go:57]: Ah, great. Thanks.

File: pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go (hunk 12 above)
Comment (timuthy) [reply pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go]: I was switching to the fakeClient when I noticed a mock is more appropriate here even if the primary goal is not to verify the calls. With mocks it's easily possible to specify or configure the behavior, especially for errors apart from `NotFound`. But as you said, not verifying the calls was awkward, so I'm now calling `ctrl.Finish()` at the end of every test case.

**Review by ialidzhikov — APPROVED** [review]
/lgtm

File: pkg/admissioncontroller/server/handlers/webhooks/utils_test.go [hunk 13]
```diff
@@ -100,10 +101,10 @@ var _ = Describe("Utils tests", func() {
 	})
 })
 
-func createHTTPRequest(obj runtime.Object, scheme *runtime.Scheme, user authenticationv1.UserInfo) *http.Request {
+func createHTTPRequest(obj runtime.Object, scheme *runtime.Scheme, user authenticationv1.UserInfo, op admissionv1beta1.Operation) *http.Request {
 	admissionReview := &admissionv1beta1.AdmissionReview{}
 
-	if obj != nil {
+	if obj != nil && !reflect.ValueOf(obj).IsNil() {
```
Comment (timebertt) [inline pkg/admissioncontroller/server/handlers/webhooks/utils_test.go:107]: What's the reason for this change?

**Review by timebertt — APPROVED** [review]
Thanks!
/lgtm

File: pkg/admissioncontroller/server/handlers/webhooks/utils_test.go (hunk 13 above)
Comment (timuthy) [reply pkg/admissioncontroller/server/handlers/webhooks/utils_test.go:107]: This change enables the following cases at the same time:
```
createHTTPRequest(nil, ...
```

and

```
var obj runtime.Object
obj = (*corev1.ConfigMap)(nil)
createHTTPRequest(obj, ...
```

Comment (timebertt) [reply pkg/admissioncontroller/server/handlers/webhooks/utils_test.go:107]: Got it, thanks :)
